### PR TITLE
Change kubevirt_vmi_storage_times_ms_total to seconds

### LINKED ISSUE #4411: [Metrics] The unit of some block metrics seems not accurate
I notice the metrics `kubevirt_vmi_storage_times_ms_total ` value is a very big number, then I check the doc of  libvirt.org, find the description which the time unit is `ns`:
```
  "block.<num>.rd.times" - total time (ns) spent on reads as
                           unsigned long long.
```
https://libvirt.org/html/libvirt-libvirt-domain.html#VIR_DOMAIN_STATS_BLOCK

### PR DESCRIPTION
**Pull request by ArthurSens**
**What this PR does / why we need it**:
This PR changes the unit of kubevirt_vmi_storage_times_ms_total to seconds, following [Prometheus best practices](https://prometheus.io/docs/practices/naming/#base-units) about time units.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4411 

**Special notes for your reviewer**:
The metric was already on an incorrect unit, following libvirt's documentation: https://libvirt.org/html/libvirt-libvirt-domain.html#VIR_DOMAIN_STATS_BLOCK
```
"block.<num>.rd.times" - total time (ns) spent on reads as unsigned long long.
"block.<num>.wr.times" - total time (ns) spent on writes as unsigned long long.
```

**Release note**:

```release-note
action required: Fix the kubevirt_vmi_storage_times_ms_total metric (reported wrong values) and rename it to kubevirt_vmi_storage_times_seconds_total. May break already existing dashboards that use this metric.
```

### CODE REVIEW

**Comment by rmohr** [comment]
`ns` seems really very drastic. I would not mind switching to seconds, if we have to change it anyway.

**Comment by ArthurSens** [comment]
> `ns` seems really very drastic. I would not mind switching to seconds, if we have to change it anyway.

I thought a little bit more about this, we wouldn't lose precision when moving to `seconds` since values are floats... We would just have a lot of decimals. Seconds it is, then

File: docs/metrics.md [hunk 1]
```diff
@@ -90,8 +90,8 @@ Extra labels:
 * `drive` - Disk device that is being written/read.
 * `type` - Whether it's a read or write operation.
 
-#### kubevirt_vmi_storage_times_ms_total
-#### HELP kubevirt_vmi_storage_times_ms_total storage operation time.
+#### kubevirt_vmi_storage_times_ns_total
```
Comment (xpivarc) [inline docs/metrics.md:93]: Does this break the signature of the metric so dashboards can be effected?

**Comment by xpivarc** [comment]
Please run `make generate` and commit the changes.

File: docs/metrics.md (hunk 1 above)
Comment (ArthurSens) [reply docs/metrics.md:93]: Yes, you're right. Forgot to mention that on release notes

**Comment by ArthurSens** [comment]
> Please run `make generate` and commit the changes.

I can't run `make generate` too, the same problem I've mentioned on slack with cluster-sync :disappointed:

**Comment by ArthurSens** [comment]
I've just run `make generate`, but looks like it isn't working as expected... Would tests pass if I change the docs manually?

**Comment by ArthurSens** [comment]
> I've just run `make generate`, but looks like it isn't working as expected... Would tests pass if I change the docs manually?

Another thing that came to my mind, some [newer versions of grafana+prometheus](https://grafana.com/blog/2020/06/15/how-we-made-working-with-prometheus-easier-with-metric-metadata-in-grafanas-explore-view/) already display metric metadata(type and Help text) on grafana's UI. Maybe the docs/metrics.md file isn't necessary anymore? Just need to improve the help texts

image

**Comment by rmohr** [comment]
> Another thing that came to my mind, some [newer versions of grafana+prometheus](https://grafana.com/blog/2020/06/15/how-we-made-working-with-prometheus-easier-with-metric-metadata-in-grafanas-explore-view/) already display metric metadata(type and Help text) on grafana's UI. Maybe the docs/metrics.md file isn't necessary anymore? Just need to improve the help texts

Sounds nice :+1:

File: docs/metrics.md (hunk 1 above)
Comment (rmohr) [inline docs/metrics.md:93]: oh wait, that should be `kubevirt_vmi_storage_times_seconds_total`.

**Review by rmohr — CHANGES_REQUESTED** [review]

File: docs/metrics.md (hunk 1 above)
Comment (ArthurSens) [reply docs/metrics.md:93]: Yep, that's what I was talking about `make generate` not working as expected...

It is not changing VMI metrics, only adding them at the end of the file as "Other metrics".
I'll just change manually for now

Comment (ArthurSens) [reply docs/metrics.md]: Should be go to go now

Comment (ArthurSens) [reply docs/metrics.md]: Oh, looks like pull_kubevirt_generate fails if I change ir manually :(

Comment (ArthurSens) [reply docs/metrics.md]: Should be good to go now

**Comment by ArthurSens** [comment]
anything else I should be doing here? @rmohr @xpivarc

**Comment by xpivarc** [comment]
> anything else I should be doing here? @rmohr @xpivarc

Maybe we want to handle this more gracefully but I am not sure about our policy. I would suggest to fix the `ms` metric and add the `seconds` if `ms` can  be seen as drastic. 

I will also answer question regarding the `metrics.md` and dashboard. While it is a very nice feature I am pesimistic about users which don't use up-to-date software and also use some other software than `Grafana`. In addition, the document serve purpouse of "offline" knowlage of these metrics so maybe someone, who is supporting customer, can refresh their memory without spinning up cluster with `Grafana`.

I am all for improving this help texts.

**Comment by rmohr** [comment]
> Maybe we want to handle this more gracefully but I am not sure about our policy. I would suggest to fix the `ms` metric and add the `seconds` if `ms` can be seen as drastic.

Yeah @xpivarc may be right. Let's fix it (and deprecate it) and just add the seconds metric. My main reasoning to still go for seconds is that I don't think that it has any practical value to report ms. It just pretends to have some use which it does not. Do others see that differently?

**Comment by ArthurSens** [comment]
> Maybe we want to handle this more gracefully but I am not sure about our policy. I would suggest to fix the `ms` metric and add the `seconds` if `ms` can be seen as drastic.

I agree that this solution is a lot more user friendly, I'll update the PR to include both metrics. Do we have a policy about how many more releases we should be keeping it?

> I will also answer question regarding the `metrics.md` and dashboard. While it is a very nice feature I am pesimistic about users which don't use up-to-date software and also use some other software than `Grafana`. In addition, the document serve purpouse of "offline" knowlage of these metrics so maybe someone, who is supporting customer, can refresh their memory without spinning up cluster with `Grafana`.

Also some pretty valid points, I now agree that this doc is still useful. The automation doesn't look to be working as expected though... You can see that `make generate` adds new metrics to the end of the file as `other metrics`, even though they are VMI metrics. Is this the intended behavior?

> I am all for improving this help texts.

I'll be doing so in another PR

File: docs/metrics.md [hunk 2]
```diff
@@ -90,8 +90,9 @@ Extra labels:
...
+#### HELP kubevirt_vmi_storage_times_ns_total storage operation time. (deprecated in favor of kubevirt_vmi_storage_times_seconds_total)
+#### HELP kubevirt_vmi_storage_times_seconds_total storage operation time.
```
Comment (ArthurSens) [inline docs/metrics.md:95]: I'm having troubles with `make generate` again, this was the result of keeping both metrics. Even though they are indeed two separate metrics. 

Result of hitting the metrics endpoint:
```
# HELP kubevirt_vmi_storage_times_ns_total storage operation time. (deprecated in favor of kubevirt_vmi_storage_times_seconds_total)
# TYPE kubevirt_vmi_storage_times_ns_total counter
kubevirt_vmi_storage_times_ns_total{drive="vda",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="read"} 2.803209064e+09
kubevirt_vmi_storage_times_ns_total{drive="vda",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="write"} 2.042818e+07
kubevirt_vmi_storage_times_ns_total{drive="vdb",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="read"} 6.0506786e+07
kubevirt_vmi_storage_times_ns_total{drive="vdb",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="write"} 0
# HELP kubevirt_vmi_storage_times_seconds_total storage operation time.
# TYPE kubevirt_vmi_storage_times_seconds_total counter
kubevirt_vmi_storage_times_seconds_total{drive="vda",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="read"} 2.803209064
kubevirt_vmi_storage_times_seconds_total{drive="vda",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="write"} 0.02042818
kubevirt_vmi_storage_times_seconds_total{drive="vdb",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="read"} 0.060506786
kubevirt_vmi_storage_times_seconds_total{drive="vdb",kubernetes_vmi_label_kubevirt_io_nodeName="node01",kubernetes_vmi_label_special="vmi-fedora",name="vmi-fedora",namespace="default",node="node01",type="write"} 0
```

**Comment by ArthurSens** [comment]
Regarding policy around changing metrics:
https://twitter.com/bwplotka/status/1327308426796560385


Creating some mixins that build alerts/dashboards for KubeVirt does sound like a smart approach :eyes: 
I'm not familiar at all with jsonnet though, it might take some months for me to come up with something useful

**Comment by fabiand** [comment]
@sradco @yuvalturg @rmohr what should we do about this PR - drop or keep and rebase?

**Comment by sradco** [comment]
@yuvalturg @andreyod @rmohr  This fixes a bug. The data is in ns but metric name says its in ms. I agree we should fix and update to seconds based Prometheus best practices. Can you please review and see of we can get this merged?

**Comment by sradco** [comment]
Closing. This issue is addressed in https://github.com/kubevirt/kubevirt/pull/5204. 
We will keep the unit as ms to keep the granularity, but fix the metric to report the correct value.

**Comment by dhiller** [comment]
/close

as requested

**Pull request closed without merging**